### ISSUE TITLE
Exponential blowup checking `dict()` of a `TypedDict` union against an unparameterized `dict` (0.0.70 regression)

### ISSUE THREAD

**Issue by aviau** [astral-sh/ty#4231]
### Intro

> (I used AI to pinpoint the issue in my repo, create this repro generator, and open this issue. I validated the finding and I think its high quality)

So today I updated `ty` from 0.0.69 to 0.0.70 and started running it against my 9k files python repo. After about 3 seconds of waiting for ty to finish typechecking, I started feeling something wasn't quite right. After 4 seconds, I knew for sure something was wrong. It didn't make sense that typechecking 9k files would take more than 4 seconds! So I (ok maybe also Claude) started debugging...

### Summary

Checking `dict(u)` against a parameter annotated as bare `dict`, where `u` is a union of N `TypedDict`s discriminated by `Literal` types, is exponential in N on ty 0.0.70 — about **2.3× per additional union member**. ty 0.0.69 checks the same code in constant time.

This looks close to #3625 and #4176, but both of those repros still pass instantly on 0.0.70. The path that survives needs one extra ingredient: the result has to be checked against an *unparameterized* `dict`. Discarding the `dict(u)` result is fast.

I hit this on a real 42-member union, where `ty check` hangs indefinitely at 100% CPU on a single file.

### Reproduction

The shape, shown at N=4:

```python
from typing import Literal
from typing import TypedDict


class D0(TypedDict):
    tag: Literal["t0"]
    field0: str


# ... D1, D2, D3 ...


U = (
    D0
    | D1
    | D2
    | D3
)


def takes_dict(d: dict) -> None: ...


def f(u: U) -> None:
    takes_dict(dict(u))
```

Generator, to scale N:

```python
# repro_gen.py — writes the repro to stdout
import sys

N = int(sys.argv[1]) if len(sys.argv) > 1 else 16

lines = ["from typing import Literal", "from typing import TypedDict", "", ""]
for i in range(N):
    lines += [
        f"class D{i}(TypedDict):",
        f'    tag: Literal["t{i}"]',
        f"    field{i}: str",
        "",
        "",
    ]
union = "\n    | ".join(f"D{i}" for i in range(N))
lines += [f"U = (\n    {union}\n)", "", ""]
lines += ["def takes_dict(d: dict) -> None: ...", "", ""]
lines += ["def f(u: U) -> None:", "    takes_dict(dict(u))", ""]

sys.stdout.write("\n".join(lines))
```

```console
$ python3 repro_gen.py 16 > repro.py
$ uvx ty@0.0.70 check --python-version 3.13 repro.py     # 39s
$ uvx ty@0.0.69 check --python-version 3.13 repro.py     # All checks passed! (0.07s)
```

### Behavior

`uvx ty@0.0.70 check --python-version 3.13 repro.py`, macOS arm64:

| N  | 0.0.70 |
|----|--------|
| 9  | 0.15s  |
| 10 | 0.27s  |
| 11 | 0.58s  |
| 12 | 1.30s  |
| 13 | 3.02s  |
| 14 | 7.10s  |
| 15 | 16.6s  |
| 16 | 38.9s  |
| 17 | 91.0s  |

There is no threshold — growth is smooth from small N. I stopped measuring at 17; extrapolating the same 2.3× factor puts N=20 around 20 minutes, and the 42-member union from my codebase never completes.

For contrast, 0.0.69 is flat: 0.06s at N=15, 0.08s at N=20, 0.07s at N=42.

### Ingredient isolation

All variants below are at N=15, where the unmodified repro takes 16.6s. Each of them completes in 0.06–0.07s:

| Variant | Result on 0.0.70 |
|---|---|
| Unmodified repro | **16.6s** |
| Drop the `dict()` call — `takes_dict(u)` | 0.06s (1 diagnostic) |
| Drop the parameter check — `x = dict(u); print(x)` | 0.07s (passes) |
| Replace `Literal["tN"]` with plain `str` | 0.07s (passes) |
| Parameterize the target — `d: dict[str, str]` | 0.07s (1 diagnostic) |

So three things are required together:

1. a union of `TypedDict`s discriminated by distinct `Literal` types,
2. a `dict(...)` constructor call over that union,
3. the result checked against an **unparameterized** `dict`.

The variants that produce a diagnostic bail out fast — the pathological case is the one where the check *passes*.

### Relationship to #3625 and #4176

Both are closed as fixed, and both stay fixed on 0.0.70, so this is not a reopen of either:

| Issue | Repro on 0.0.70 |
|---|---|
| #3625 — `dict()` on a discriminated `TypedDict` union hangs type checking | `All checks passed!` |
| #4176 — Exponentially slow type check time with large `TypedDict` union | `All checks passed!` in 0.06s |

The diagnosis in #4176 — "probably creating a large constraint set for the `K` and `V` type variables of the returned `dict[K, V]` type" — may apply here too: discarding the result is cheap, but checking it against an unparameterized `dict` appears to force those constraints to be solved.

### Version

ty 0.0.70 (bf8c9f3de 2026-08-10)

macOS 25.6.0 arm64, `--python-version 3.13`, no config file, no third-party dependencies.

**Comment by carljm** [comment]
Thanks for the report! Confirmed, can reproduce. Looks like the regression was introduced by https://github.com/astral-sh/ruff/pull/27614, cc @dcreager .

**Comment by dcreager** [comment]
I'll take a look!

**Comment by dcreager** [comment]
https://github.com/astral-sh/ruff/pull/27655 should fix this
